discord/commands/mixins.py

```diff
@@ -0,0 +1,952 @@
+"""
+The MIT License (MIT)
+
+Copyright (c) 2015-2021 Rapptz
+Copyright (c) 2021-present Pycord Development
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import datetime
+import functools
+from typing import TYPE_CHECKING, Any, Callable, Coroutine, Generic, TypeVar, Union
+
+from .. import abc, utils
+from ..errors import (
+    ApplicationCommandError,
+    CheckFailure,
+    CommandError,
+    CommandInvokeError,
+    CommandOnCooldown,
+    DisabledCommand,
+)
+from .cooldowns import BucketType, Cooldown, CooldownMapping, MaxConcurrency
+
+T = TypeVar("T")
+BotT = TypeVar("BotT", bound="Union[Bot, AutoShardedBot]")
+CogT = TypeVar("CogT", bound="Cog")
+
+Coro = Coroutine[Any, Any, T]
+MaybeCoro = Union[T, Coro[T]]
+
+Check = Union[
+    Callable[[CogT, "BaseContext"], MaybeCoro[bool]],
+    Callable[["BaseContext"], MaybeCoro[bool]],
+]
+
+Error = Union[
+    Callable[[CogT, "BaseContext[Any]", CommandError], Coro[Any]],
+    Callable[["BaseContext[Any]", CommandError], Coro[Any]],
+]
+ErrorT = TypeVar("ErrorT", bound="Error")
+
+Hook = Union[
+    Callable[[CogT, "BaseContext"], Coro[Any]], Callable[["BaseContext"], Coro[Any]]
+]
+HookT = TypeVar("HookT", bound="Hook")
+
+if TYPE_CHECKING:
+    from typing_extensions import Concatenate, ParamSpec
+
+    from ..abc import MessageableChannel
+    from ..bot import AutoShardedBot, Bot
+    from ..cog import Cog
+    from ..guild import Guild
+    from ..interactions import Interaction
+    from ..member import Member
+    from ..message import Message
+    from ..state import ConnectionState
+    from ..user import ClientUser, User
+    from ..voice_client import VoiceProtocol
+
+    P = ParamSpec("P")
+
+    Callback = (
+        Callable[Concatenate[CogT, "BaseContext", P], Coro[T]]
+        | Callable[Concatenate["BaseContext", P], Coro[T]]
+    )
+else:
+    P = TypeVar("P")
```

Comment:
Why is this defined as a `TypeVar` at runtime? What's wrong with using `ParamSpec` from typing_extensions at runtime?